zuko/flows/autoregressive.py

```diff
@@ -102,33 +107,87 @@ def __init__(
 
         if passes is None:
             passes = features
+        self.passes = min(max(passes, 1), features)
+
+        assert (order is None) or (
+            adjacency is None
+        ), "Parameters `order` and `adjacency_matrix` are mutually exclusive."
+        assert (passes == features) or (
+            adjacency is None
+        ), "When using `adjacency_matrix`, `passes` has to be the number of features."
+
+        if adjacency is None:
+            if order is None:
+                order = torch.arange(features)
+            else:
+                order = torch.as_tensor(order)
+
+            self.order = torch.div(order, ceil(features / self.passes), rounding_mode='floor')
 
-        if order is None:
-            order = torch.arange(features)
+            in_order = torch.cat((self.order, torch.full((context,), -1)))
+            out_order = torch.repeat_interleave(self.order, self.total)
+            adjacency = out_order[:, None] > in_order
         else:
-            order = torch.as_tensor(order)
+            adjacency.mul_(  # Remove the diagonal (if it is non-zero)
```

Comment:
I opted for doing it silently since "in reality" the diagonal is non-zero (because of the transformer), and it needs to be made zero for the conditioner. I figured it out that fixing it would be less confusing, but I can also just raise an error.

Reply:
I see what you mean, but then we should impose that the matrix has ones on the diagonal, and raise a error if it has not.

The reason why I prefer raising an error is that if a user makes a mistake (e.g. they permute the rows but they forget to permute the columns), they might not see it if we "fix" it silently.

Reply:
That is actually a pretty good reason to checking it. I'll raise an error then 👍🏼